src/dns.c

```diff
@@ -431,13 +469,13 @@ hsk_dns_msg_read(uint8_t **data, size_t *data_len, hsk_dns_msg_t *msg) {
       msg->edns.rd = opt->rd;
       msg->code |= msg->edns.code << 4;
 
+      hsk_dns_rr_t *popped = hsk_dns_rrs_pop(&msg->ar);
+      assert(popped == rr);
       free(rr->rd);
```

Comment:
`hsk_dns_rr_free` would actually free too much and I tried it, and ended up with `pointer being freed was not allocated` thrown from `hsk_dns_msg_uninit(msg=0x0000000103009000) at dns.c:99:5` which is the big uninit at the very end of that object's life. The reason is because we still want to save some of the data from the original message:

```
      msg->edns.rd_len = opt->rd_len;
      msg->edns.rd = opt->rd;
```

If we `hsk_dns_rr_free()` this memory would be freed, and then we'll try to free that memory again later. The OPT record is a bit funny, especially the variable naming here, because it has its own record data or `rd` member.

So we have:

```
typedef struct hsk_dns_rr_s {
  uint8_t name[HSK_DNS_MAX_NAME];
  uint16_t type;
  uint16_t class;
  uint32_t ttl;
  void *rd; --->>> which for OPT will be hsk_dns_opt_rd_t...
} hsk_dns_rr_t;

typedef struct {
  size_t rd_len;
  uint8_t *rd; --->>> this data is copied to msg.edns.rd and freed by hsk_dns_msg_uninit()
} hsk_dns_opt_rd_t;
```


essentially `rr->rd->rd` 🙄